cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -225,6 +225,423 @@ TEST(TestScalarNested, StructField) {
+        ],
+        [
+          "values",
+          [[6, 6], [7, 7]]
+        ],
+        [
+          "foo",
+          [[8, 8], [9, 9]]
+        ],
+        [
+          "point",
+          [[10, 10], [11, 11]]
+        ],
+        [
+          "foo",
+          [[12, 12], [13, 13]]
+        ]
+      ],
+      null,
+      [
+        [
+          "yet",
+          [[0, 1], [1, 2]]
+        ],
+        [
+          "more",
+          [[2, 3], [3, 4]]
+        ],
+        [
+          "foo",
+          [[4, 5], [5, 6]]
+        ],
+        [
+          "random",
+          [[6, 7], [7, 8]]
+        ],
+        [
+          "foo",
+          [[8, 9], [9, 10]]
+        ],
+        [
+          "values",
+          [[10, 11], [11, 12]]
+        ],
+        [
+          "foo",
+          [[12, 13], [13, 14]]
+        ]
+      ],
+      []
+    ]
+  )";
+  const auto map_array = ArrayFromJSON(type, input);
+
+  const auto expected_all = ArrayFromJSON(list(map(int16(), int16())), R"(
+                                [
+                                  [
+                                    [[4, 4], [5, 5]], [[8, 8], [9, 9]],
+                                    [[12, 12], [13, 13]]
+                                  ],
+                                  null,
+                                  [
+                                    [[4, 5], [5, 6]], [[8, 9], [9, 10]],
+                                    [[12, 13], [13, 14]]
+                                  ],
+                                  null
+                                ])");
+  const auto expected_first = ArrayFromJSON(map(int16(), int16()), R"(
+                                [
+                                  [[4, 4], [5, 5]],
+                                  null,
+                                  [[4, 5], [5, 6]],
+                                  null
+                                ])");
+  const auto expected_last = ArrayFromJSON(map(int16(), int16()), R"(
+                                [
+                                  [[12, 12], [13, 13]],
+                                  null,
+                                  [[12, 13], [13, 14]],
+                                  null
+                                ])");
+
+  CheckMapArrayLookupWithDifferentOptions(map_array, MakeScalar("foo"), expected_all,
+                                          expected_first, expected_last);
+}
+
+TEST_F(TestMapArrayLookupKernel, BooleanKey) {
+  auto true_scalar = ScalarFromJSON(boolean(), R"(true)");
+  auto map_type = map(boolean(), int32());
+  const char* input = R"(
+    [
+      [
+        [true, 99], [false, 1], [false, 2], [true, null], [false, 5],
+        [true, 8]
+      ],
+      null,
+      [
+        [false, null], [true, 67], [false, 101], [false, 1], [false, null],
+        [false, 9], [true, 80]
+      ],
+      [],
+      [
+        [false, 1], [false, 2], [false, 3], [false, 4]
+      ],
+      [
+        [true, 9], [true, 2], [true, 5], [true, 8]
```

Comment:
In general: can we add some `null` keys in addition to `true` and `false`? (Same below.)

Reply:
This causes error: https://github.com/apache/arrow/blob/a3efe72f99b1b9f23b1d11afc648f4306d32e330/cpp/src/arrow/ipc/json_simple.cc#L585-L586

Reply:
Ah, I didn't realize that was disallowed. Ok then, we should be good. And that means we don't have to decide what to do with null query keys. That also means we should check if the query key is null and return an error in that case.